### PR TITLE
[WFLY-13632] If a CMR resource is deployed concurrently then it will overwrite the list of resources that are expected to behave as CMR

### PR DESCRIPTION
**Pull request by gaol**
~Issue: https://issues.redhat.com/browse/WFLY-13633~

Issue: https://issues.redhat.com/browse/WFLY-13632

Thanks for submitting your Pull Request!

Please delete this text, and add a link to the Jira issue solved by this PR.

Remember to use the Jira issue ID in the PR title, and any commits.

### CODE REVIEW

**Comment by gaol** [comment]
@tomjenkinson @bstansberry @mmusgrov Could anyone please review? Thanks :)

**Comment by bstansberry** [comment]
This LGTM but let's wait a bit in case @mmusgrov has any concerns.

**Comment by gaol** [comment]
@tomjenkinson doen't [jtaBean.setCommitMarkableResourceJNDINames(connectableResourceJNDINames);](https://github.com/wildfly/wildfly/pull/13525/files#diff-104e9239790ebc24b14cf99ab71fc053R94) update the underlying collection in narayana ?

**Comment by gaol** [comment]
@tomjenkinson OK, I understand now. Thank you. It is more than I expected, and it needs changes in Narayana first to keep all consistent.

**Comment by mmusgrov** [comment]
Yes Tom is correct and if we cannot perform a server restart then we would also need a fix in Narayana.

**Comment by mmusgrov** [comment]
@gaol Hi, I created a [JBTM issue](https://issues.redhat.com/browse/JBTM-3357) to ensure that these property getters are dynamic.

**Comment by bstansberry** [comment]
@gaol [~gaol]. Note that in WildFly CMResourceAdd has a bug in how it deals with reload-required:

https://github.com/wildfly/wildfly/blob/master/transactions/src/main/java/org/jboss/as/txn/subsystem/CMResourceAdd.java#L80

That entire method should really be an if/else block -- it either the server isn't booting and it just calls context.setReloadRequired() or it is booting and it modifies the runtime state by adding the service.

Right now my *impression* from the discussion is that doesn't matter too much because if CMResourceAdd runs after boot, the service it adds doesn't actually update the state that Narayana is really using. So it's doing pointless work, but it's not modifying the actual runtime behavior and then falsely telling the user it didn't by setting reload-required.

That should be fixed regardless, but particularly if things change in how Narayana handles updates made to JTAEnvironmentBean. Basically, what the handlers for the 'add' and 'remove' ops on these resources do should match what happens inside Narayana. If a change doesn't take effect without a process restart, then they should set restart-required. If only a reload is needed, then reload-required. If they are able immediately to correctly change what Narayana is doing, then no reload-required or restart-required.

The synchronization stuff here is good regardless, because during boot multiple CMResourceService instances can be installed, and they need to coordinate.

**Comment by gaol** [comment]
@bstansberry Thanks for the review :)  

Here are summary of the changes:

* I updated the `CMResourceAdd` part to address the issue Brian mentioned.
* I updated the PR summary and description to target https://issues.redhat.com/browse/WFLY-13632 only.
* The https://github.com/jbosstm/narayana/pull/1670 may solve https://issues.redhat.com/browse/WFLY-13633 together with this fix.

This PR doesn't fix https://issues.redhat.com/browse/WFLY-13633 because updates to `JTAEnvironmentBean` by `CMResourceService` does not change the **static references** to `JTAEnvironmentBean.commitMarkableResourceJNDINames`([for example](https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L1748)) in other places in narayana.

After `:reload` operation, the static references to `JTAEnvironmentBean.commitMarkableResourceJNDINames`([for example](https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L1748)) are still the same because they are cached as the static class variables.

Together with https://github.com/jbosstm/narayana/pull/1670, which changes the static class fields to final instance fields, the `:reload` operation should be able to re create all needed instances with updated `JTAEnvironmentBean`.

But I think setting the state to `restart-required` should also be able to fix the problem, I am not sure if changes to `JTAEnvironmentBean` should lead to a hard restart or not.

**Comment by mmusgrov** [comment]
This PR will protect against the data integrity issue if we mark the server as `restart-required`.
@gaol would you be able to update your PR to include that change?

File: transactions/src/main/java/org/jboss/as/txn/subsystem/CMResourceAdd.java [hunk 1]
```diff
@@ -64,6 +64,10 @@ protected void populateModel(ModelNode operation, ModelNode model) throws Operat
      */
     @Override
     protected void performRuntime(final OperationContext context, final ModelNode operation, final ModelNode model) throws OperationFailedException {
+        if (!context.isBooting()) {
+            context.reloadRequired();
```
Comment (mmusgrov) [inline transactions/src/main/java/org/jboss/as/txn/subsystem/CMResourceAdd.java:68]: Can we change this to restart to force Narayana to reread its config?

**Comment by bstansberry** [comment]
Superseded by #13620

**Pull request closed without merging**